compiler/rustc_errors/src/lib.rs

```diff
@@ -451,6 +451,14 @@ struct HandlerInner {
     /// If set to `true`, no warning or error will be emitted.
     quiet: bool,
 
+    /// Expected [`Diagnostic`]s store a [`LintExpectationId`] as part of
+    /// the lint level. [`LintExpectationId`]s created early during the compilation
+    /// (before `HirId`s have been defined) are not stable and can therefore not be
+    /// stored on disk. This buffer stores these diagnostics until the ID has been
+    /// replaced by a stable [`LintExpectationId`]. The [`Diagnostic`]s are the
+    /// submitted for storage and added to the list of fulfilled expectations.
+    unstable_expect_diagnostics: Vec<Diagnostic>,
```

Comment:
Perhaps we should assert this is empty in `steal_fulfilled_expectation_ids`?

Reply:
That's probably a good idea :+1: